// @tencent/ast-string-extractor/src/utils/file-langs.js
const path = require('path');

// 语言类型常量
const LANG_TYPES = {
    TS: 'ts',
    JS: 'js',
    HTML: 'html',
    VUE: 'vue',
    XML: 'xml'
};

// 后缀名到语言类型的映射
const EXTENSION_MAP = {
    '.tsx': LANG_TYPES.TS,
    '.ts': LANG_TYPES.TS,
    '.jsx': LANG_TYPES.JS,
    '.js': LANG_TYPES.JS,
    '.html': LANG_TYPES.HTML,
    '.htm': LANG_TYPES.HTML,
    '.vue': LANG_TYPES.VUE,
    '.xml': LANG_TYPES.XML,
};

function isJsFile(filePath) {
    return detectLanguage(filePath) === LANG_TYPES.JS;
}

function isTsFile(filePath) {
    return detectLanguage(filePath) === LANG_TYPES.TS;
}

function isHtmlFile(filePath) {
    return detectLanguage(filePath) === LANG_TYPES.HTML;
}

function isVueFile(filePath) {
    return detectLanguage(filePath) === LANG_TYPES.VUE;
}

function isXmlFile(filePath) {
    return detectLanguage(filePath) === LANG_TYPES.XML;
}

/**
 * 根据文件路径识别语言类型
 * @param {string} filePath - 文件路径
 * @returns {string|null} 返回语言类型字符串，无法识别时返回null
 */
function detectLanguage(filePath) {
    const ext = path.extname(filePath).toLowerCase();
    return EXTENSION_MAP[ext] || null;
}

module.exports = {
    LANG_TYPES,
    detectLanguage,
    isJsFile,
    isTsFile,
    isHtmlFile,
    isVueFile,
    isXmlFile
};
